python/src/iceberg/utils/schema_conversion.py

```diff
@@ -0,0 +1,451 @@
+        else:
+            avro_types = type_union
+
+        is_optional = "null" in avro_types
+
+        if len(avro_types) > 2:
+            raise TypeError("Non-optional types aren't part of the Iceberg specification")
+
+        # Filter the null value and return the type
+        return list(filter(lambda t: t != "null", avro_types))[0], is_optional
+
+    def _resolve_inner_type(self, avro_type: Dict[str, Any], inner_field_name: str, id_field: str) -> Tuple[IcebergType, bool]:
+        """
+        Removes any unions inside an array or map, and appends the field-id downstream (if required)
+
+        Args:
+            avro_type: The parent Avro type
+            inner_field_name: For example 'items' in case of an array
+            id_field: Contains the element or key/value id, for example element-id, to be copied to the field-id
+
+        Returns:
+            The inner type and the nullability
+        """
+        plain_type, element_is_optional = self._resolve_union(avro_type[inner_field_name])
+        if isinstance(plain_type, dict):
+            # We need the element-id downstream
+            plain_type["field-id"] = avro_type[id_field]
+        inner_field = self._convert_schema(plain_type)
+        return inner_field, element_is_optional
+
+    def _convert_schema(self, avro_type: str | Dict[str, Any]) -> IcebergType:
+        """
+        Resolves the Avro type
+
+        Args:
+            avro_type: The Avro type, can be simple or complex
+
+        Returns:
+            The equivalent IcebergType
+
+        Raises:
+            ValueError: When there are unknown types
+        """
+        if isinstance(avro_type, str):
+            return PRIMITIVE_FIELD_TYPE_MAPPING[avro_type]
+        elif isinstance(avro_type, dict):
+            if "logicalType" in avro_type:
+                return self._convert_logical_type(avro_type)
+            else:
+                type_identifier = avro_type["type"]
+                if type_identifier == "record":
+                    return self._convert_record_type(avro_type)
+                elif type_identifier == "array":
+                    return self._convert_array_type(avro_type)
+                elif type_identifier == "map":
+                    return self._convert_map_type(avro_type)
+                elif type_identifier == "fixed":
+                    return self._convert_fixed_type(avro_type)
+                else:
+                    raise ValueError(f"Unknown type: {avro_type}")
+        else:
+            raise ValueError(f"Unknown type: {avro_type}")
+
+    def _convert_field(self, field: Dict[str, Any]) -> NestedField:
+        """
+        Converts an Avro field into an Iceberg equivalent field
+        Args:
+            field: The Avro field
+
+        Returns:
+            The Iceberg equivalent field
+        """
+        assert "field-id" in field, "Missing field-id in the Avro field, this is required for converting it to an Iceberg schema"
```

Comment:
What is thrown when using `assert` rather than `raise ValueError`?

Also, we try to be more direct with error messages by starting with what was attempted, then giving context for what went wrong. It's also assumed that if an exception is raised then some requirement wasn't met so we'd typically leave out the "this is required ..." part. How about `f"Cannot convert field, missing field-id: {field}"`?

Reply:
Great point. I'm not against assert's personally, but there is a battle in the Python community if you should use them for checks like these. Technically you can ignore them using a flag `python -O`, and they are meant for developers. I've changed them to a ValueError to be in line with the rest of the exceptions (which I like).

I've updated the message to your suggestion. I think it is good to have the field in there 👍🏻